sdk/dynamicconfig/metric/controller/push/period_matcher.go

```diff
@@ -51,26 +52,46 @@ func (matcher *PeriodMatcher) MarkStart(startTime time.Time) {
 // ApplySchedules sets the schedules that a PeriodMatcher consults when
 // constructing a Rule. After processing the schedules, ApplySchedules returns
 // the optimal period with which a controller should run a collection sweep.
+// If a period of 0 is returned, then metric collection should be halted.
 //
 // This function may be called concurrently.
-func (matcher *PeriodMatcher) ApplySchedules(sched []*pb.MetricConfigResponse_Schedule) time.Duration {
+func (matcher *PeriodMatcher) ApplySchedules(sched []*pb.MetricConfigResponse_Schedule) (time.Duration, error) {
+	if err := validate(sched); err != nil {
+		return 0, err
+	}
+
 	matcher.m.Lock()
 	matcher.sched = sched
 	matcher.metrics = make(map[string]*collectData)
 	matcher.m.Unlock()
 
-	return getExportPeriod(matcher.sched)
+	return getExportPeriod(matcher.sched), nil
+}
+
+func validate(sched []*pb.MetricConfigResponse_Schedule) error {
+	if len(sched) == 0 {
+		return errors.New("no schedules")
+	}
+
+	for _, schedule := range sched {
+		if schedule.PeriodSec < 0 {
+			return errors.New("periods must be positive")
+		}
+	}
+
+	return nil
 }
 
-// TODO: handle explicit zeros
 func getExportPeriod(sched []*pb.MetricConfigResponse_Schedule) time.Duration {
 	if len(sched) == 0 {
 		panic("matcher has not applied any schedules")
 	}
 
-	checkPeriod := sched[0].PeriodSec
-	for _, schedule := range sched[1:] {
-		checkPeriod = gcd(checkPeriod, schedule.PeriodSec)
+	var checkPeriod int32
```

Comment:
this got initialized to 0, right? Does the test pass?

Reply:
Good catch! Updated message.

For the second point, Euclid's algorithm, given a zero and positive value, will select the positive value. Hence we keep checkPeriod uninitialized (i.e. 0) so that it can be initialized by the first positive integer in the periods. If no such integer exists (i.e. when all periods are 0), then checkPeriod remains zero, which signals the user to disable metric collection. Disabling the ticker under the current implementation is a little messy and would probably not be worth the code overhaul, so I've gone ahead and just set the period really long (1 week) in this event. When that week passes, a rule is produced that returns false for all metrics, effectively disabling collection.

Reply:
Thanks for the explanation. Let's add a comment about this. Preferably a test as well.

Reply:
Yup! Tests have been updated. The comments were elaborated in the previous commit, but I've gone ahead and put a little note on the gcd function clarifying its function.